src/GraphQL/Execution/ResolveFieldContextSlim.cs

```diff
@@ -0,0 +1,119 @@
+using GraphQL.Instrumentation;
+using GraphQL.Language.AST;
+using GraphQL.Types;
+using System;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading;
+using Field = GraphQL.Language.AST.Field;
+
+namespace GraphQL.Execution
+{
+    public static class ResolveFieldContextSlim
+    {
+        private static readonly ConcurrentDictionary<System.Type, Func<ExecutionNode, ExecutionContext, IResolveFieldContext>> dic =
+            new ConcurrentDictionary<System.Type, Func<ExecutionNode, ExecutionContext, IResolveFieldContext>>();
+
+        public static IResolveFieldContext Create(ExecutionNode node, ExecutionContext context)
+        {
+            //get the type of the source object, and just use typeof(object) when the source is null
+            var sourceType = node.Source?.GetType() ?? typeof(object);
+
+            //retrieve the constructor
+            var func = dic.GetOrAdd(sourceType, LazyCreate);
+
+            //execute the constructor and return the new ResolveFieldContextSlim<T> (cast as IResolveFieldContext)
+            return func(node, context);
+        }
+
+        private static Func<ExecutionNode, ExecutionContext, IResolveFieldContext> LazyCreate(System.Type sourceType)
+        {
+            //rfcsType = typeof(ResolveFieldContextSlim<sourceType>);
+            var rfcsType = typeof(ResolveFieldContextSlim<>).MakeGenericType(sourceType);
+
+            //constructor = (the only constructor)
+            var constructor = rfcsType.GetConstructors()[0];
+
+            //create the 2 parameters to the lambda: executionNode and executionContext
+            var executionNodeParameter = Expression.Parameter(typeof(ExecutionNode), "executionNode");
+            var executionContextParameter = Expression.Parameter(typeof(ExecutionContext), "executionContext");
+
+            //body = new ResolveFieldContextSlim<sourceType>(executionNode, executionContext);
+            var expressionBody = Expression.New(constructor, executionNodeParameter, executionContextParameter);
+
+            //lambda expression = (executionNode, executionContext) => new ResolveFieldContextSlim<sourceType>(executionNode, executionContext);
+            var lambdaExpression = Expression.Lambda<Func<ExecutionNode, ExecutionContext, IResolveFieldContext>>(expressionBody, executionNodeParameter, executionContextParameter);
+
+            //compile the expression into a function and return it
+            return lambdaExpression.Compile();
+        }
+    }
+
+    public class ResolveFieldContextSlim<T> : IResolveFieldContext<T>
+    {
+        private readonly ExecutionNode _executionNode;
+        private readonly ExecutionContext _executionContext;
+        private readonly Lazy<Dictionary<string, object>> _arguments;
```

Comment:
Do not use Lazy. Less allocations is better. There are no race conditions since `ResolveFieldContext` is used only by one resolver.

Reply:
Done